### PR TITLE
Issue #1123 datasets doc improvement

### PR DESCRIPTION
**Pull request by lulmer**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.
https://github.com/pytorch/torchtune/issues/1123, #1226 

#### Changelog
What are the changes made in this PR?

This PR adds more details for
- the datasets tutorial, especially for newcomers that are not aware that huggingface datasets handle local datasets too. 
- the api doc of `instruct_template` with suggested explanation from @ebsmothers in [issue #1099 ](https://github.com/pytorch/torchtune/issues/1099#issuecomment-2186081011)

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [x] [building and checking doc rendering](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md#building-docs)

### CODE REVIEW

File: docs/source/tutorials/datasets.rst [hunk 1]
```diff
@@ -52,15 +52,15 @@ function. You can see a list of all supported datasets :ref:`here<datasets>`.
...
+Hugging Face and Local datasets
+-------------------------------
 
 We provide first class support for datasets on the Hugging Face hub. Under the hood,
 all of our built-in datasets and dataset builders are using Hugging Face's ``load_dataset()``
 to load in your data, whether local or on the hub.
 
 You can pass in a Hugging Face dataset path to the ``source`` parameter in any of our builders
-to specify which dataset on the hub to download. Additionally, all builders accept
+to specify which dataset on the hub to download or use from a local directory path. Additionally, all builders accept
```
Comment (RdoubleA) [inline docs/source/tutorials/datasets.rst:63]: could you add a ref link to the local datasets section here? Since source has to be configured slightly differently when using local files. See [here](https://docs.readthedocs.io/en/stable/guides/cross-referencing-with-sphinx.html#implicit-targets) for how to do that with RST

File: docs/source/tutorials/datasets.rst [hunk 2]
```diff
@@ -295,6 +295,21 @@ and create your own class.
     dataset.template=import.path.to.CustomTemplate
 
 
+Under the hood, torchtune is using :code:`importlib.import_module`.
```
Comment (RdoubleA) [inline docs/source/tutorials/datasets.rst:298]:
```suggestion
Under the hood, torchtune is using :code:`importlib.import_module` to locate components from their dotpaths.
```

File: docs/source/tutorials/datasets.rst [hunk 3]
```diff
@@ -295,6 +295,21 @@ and create your own class.
     dataset.template=import.path.to.CustomTemplate
 
 
+Under the hood, torchtune is using :code:`importlib.import_module`. 
+You can place your custom template class in any Python module as long as the module is
```
Comment (RdoubleA) [inline docs/source/tutorials/datasets.rst:299]: nit: module might suggest modifying a preexisting library
```suggestion
You can place your custom template class in any Python file as long as the file is 
```

File: docs/source/tutorials/datasets.rst [hunk 4]
```diff
@@ -295,6 +295,21 @@ and create your own class.
...
 
+Under the hood, torchtune is using :code:`importlib.import_module`. 
+You can place your custom template class in any Python module as long as the module is 
+accessible by Python's import mechanism. This means the module should be in a directory 
+that is included in Python's search paths (:code:`sys.path`). 
+This often includes:
+
+- The current directory from which your Python interpreter or script is run.
+- Directories where Python packages are installed (like :code:`site-packages`).
+- Any directories added to :code:`sys.path` at runtime or through the :code:`PYTHONPATH` environment variable.
```
Comment (RdoubleA) [inline docs/source/tutorials/datasets.rst:306]:
```suggestion
- Any directories added to :code:`sys.path` at runtime using :code:`sys.path.append` or through the :code:`PYTHONPATH` environment variable.
```

File: docs/source/tutorials/datasets.rst [hunk 5]
```diff
@@ -295,6 +295,21 @@ and create your own class.
...
+
+If you are working within a cloned repository of torchtune you can just define your custom template
```
Comment (RdoubleA) [inline docs/source/tutorials/datasets.rst:308]: I know this is a viable option, but I would rather encourage users to create custom components in their own project directories instead of having to modify torchtune internals just to run fine-tuning jobs. From a library maintainer's perspective, this doesn't seem like a good user flow - they should be able to import our library and make their custom scripts and building blocks using our components - just like with core PyTorch. And if there's anything that can be improved to make that workflow easier then we need to do that. What do you think?

File: torchtune/datasets/_instruct.py [hunk 6]
```diff
@@ -46,7 +46,9 @@ class InstructDataset(Dataset):
         transform (Optional[Callable]): transform to apply to the sample before formatting to the template.
             Default is None.
         column_map (Optional[Dict[str, str]]): a mapping from the expected placeholder names in the template
-            to the column/key names in the sample. If None, assume these are identical.
+            to the column/key names in the sample. If None, assume these are identical. 
+            the output ground truth field can be provided to through the `output` key mapping.
```
Comment (RdoubleA) [inline torchtune/datasets/_instruct.py:50]:
```suggestion
            The output column can be indicated using the ``output`` key mapping. 
```

File: torchtune/datasets/_instruct.py [hunk 7]
```diff
@@ -46,7 +46,9 @@ class InstructDataset(Dataset):
         transform (Optional[Callable]): transform to apply to the sample before formatting to the template.
             Default is None.
         column_map (Optional[Dict[str, str]]): a mapping from the expected placeholder names in the template
-            to the column/key names in the sample. If None, assume these are identical.
+            to the column/key names in the sample. If None, assume these are identical. 
+            the output ground truth field can be provided to through the `output` key mapping. 
+            If no placeholder for the `output` field is provided in `column_map` it is assumed to be `output`.
```
Comment (RdoubleA) [inline torchtune/datasets/_instruct.py:51]:
```suggestion
            If no placeholder for the ``output`` column is provided in ``column_map`` it is assumed to be ``output``.
```

**Review by RdoubleA — COMMENTED** [review]
This is an awesome update - you really found the blind spot in our documentation but is still super important. Honestly, this new section may need to be placed somewhere more visible, but this is not blocking for this PR.

**Comment by ebsmothers** [comment]
Pinging this PR since it's been quiet for a bit. @lulmer do you have the bandwidth to address the comments? If not, maybe @RdoubleA can help take this over the finish line.

**Comment by lulmer** [comment]
Woops sorry, was preparing another (huge PR), I will do something about it today

File: docs/source/tutorials/datasets.rst (hunk 5 above)
Comment (lulmer) [reply docs/source/tutorials/datasets.rst:308]: I agree, modifying library internals should be only for maintainers and contributors. 
Should we phrase it like : create your own custom dir that contains your configs and custom components you want to link to torchtune ? 
But this would deserve I whole walkthrough because there are different level of complexity : from the user that just as a custom dataset from an user that has a whole new model or loss function

File: docs/source/tutorials/datasets.rst (hunk 1 above)
Comment (lulmer) [reply docs/source/tutorials/datasets.rst:63]: I am not sure to understand where the ref link is supposed to point to. Can you be more explicit ?

Comment (pbontrager) [reply docs/source/tutorials/datasets.rst:63]: https://pytorch.org/torchtune/stable/tutorials/datasets.html#local-and-remote-datasets

File: docs/source/tutorials/datasets.rst (hunk 5 above)
Comment (pbontrager) [reply docs/source/tutorials/datasets.rst:308]: You could say "In your own repository, define your custom template in its own file and you can then access directly from your config by providing the import path"

File: docs/source/tutorials/datasets.rst (hunk 1 above)
Comment (RdoubleA) [reply docs/source/tutorials/datasets.rst:63]: sorry, I meant to add it like this:
```suggestion
to specify which dataset on the hub to download or use from a local directory path (see `Local and remote datasets`_). Additionally, all builders accept
```

File: docs/source/tutorials/datasets.rst [hunk 8]
```diff
@@ -52,15 +52,15 @@ function. You can see a list of all supported datasets :ref:`here<datasets>`.
     tune run full_finetune_single_device --config llama3/8B_full_single_device \
     dataset=torchtune.datasets.alpaca_dataset
 
-Hugging Face datasets
----------------------
+Hugging Face and Local datasets
```
Comment (RdoubleA) [inline docs/source/tutorials/datasets.rst:55]: Local is already covered in a different section, let's remove it from here to avoid confusion?

**Comment by RdoubleA** [comment]
@lulmer sorry I got to this so late, let me know if you'd like me to finish this up or if you're still open to addressing the comments. I'd like to get this merged in

**Review by ebsmothers — APPROVED** [review]

File: docs/source/tutorials/datasets.rst [hunk 9]
```diff
@@ -295,6 +295,17 @@ and create your own class.
     dataset.template=import.path.to.CustomTemplate
 
 
+Under the hood, torchtune is using :code:`importlib.import_module` to locate components from their dotpaths.
```
Comment (joecummings) [inline docs/source/tutorials/datasets.rst:298]: Nit: we already use "under the hood" a couple times in this tutorial.

Comment (joecummings) [reply docs/source/tutorials/datasets.rst:298]: Also can we link to a reference about import_module?

**Review by joecummings — APPROVED** [review]
Couple of comments that'd I'd love to see addressed but overall LGTM

**Comment by RdoubleA** [comment]
@lulmer Thanks for your help, I went ahead and cleaned this up to get this merged. Let me know if you have any remaining concerns
